internal/validator/validate_test.go

```diff
@@ -0,0 +1,81 @@
+package validator
+
+import (
+	"bytes"
+	"io/fs"
+	"os"
+	"path/filepath"
+	"strings"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"golang.org/x/tools/txtar"
+
+	// Install Components
+	_ "github.com/grafana/alloy/internal/component/all"
+
+	"github.com/grafana/alloy/internal/component"
+	"github.com/grafana/alloy/internal/featuregate"
+)
+
+const (
+	txtarSuffix = ".txtar"
+	diagsSuffix = ".diags"
+)
+
+func TestValidate(t *testing.T) {
+	// Test with default config.
+	testDirectory(t, "./testdata/default", featuregate.StabilityGenerallyAvailable, false)
+}
+
+func testDirectory(t *testing.T, dir string, minStability featuregate.Stability, enableCommunityComps bool) {
+	require.NoError(t, filepath.WalkDir(dir, func(path string, d fs.DirEntry, _ error) error {
+		if d.IsDir() && path != dir {
+			return filepath.SkipDir
+		}
+
+		if strings.HasSuffix(path, txtarSuffix) {
+			archive, err := txtar.ParseFile(path)
+			require.NoError(t, err)
+
+			t.Run(strings.TrimSpace(string(archive.Comment)), func(t *testing.T) {
+				sources := make(map[string][]byte, len(archive.Files))
+				for _, f := range archive.Files {
+					sources[f.Name] = f.Data
+				}
+
+				validateErr := Validate(Options{
+					Sources:           sources,
+					ComponentRegistry: component.NewDefaultRegistry(minStability, enableCommunityComps),
+				})
+
+				diagsFile := strings.TrimSuffix(path, txtarSuffix) + diagsSuffix
+				if !fileExists(diagsFile) {
```

Comment:
Currently the test will create a snapshot file if it don't exists. So to update one we just have to delete it and run tests again.

We can look into other approaches to this. https://github.com/gkampitakis/go-snaps?tab=readme-ov-file#update-snapshots could be an option. We create / update files if  `UPDATE_SNAPSHOT=true` env is set

Reply:
That works, but relies a bit on a convention that may be hard to find / understand. 

I like the approach we used here, where test has a flag: https://github.com/grafana/alloy/blob/main/internal/tools/docs_generator/docs_updated_test.go#L23 

...if it's possible to adopt it here would be great, but I don't mind too much.

Reply:
I think its better to be consistent. So I will add it for these tests too

Reply:
Something like this [81a50d3](https://github.com/grafana/alloy/pull/3220/commits/81a50d3d01d7a45438db4be8addbd9baa477992a)